src/io/flutter/actions/FlutterRunModeAction.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2017 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+package io.flutter.actions;
+
+import com.intellij.execution.Executor;
+import com.intellij.execution.ProgramRunnerUtil;
+import com.intellij.execution.RunManagerEx;
+import com.intellij.execution.RunnerAndConfigurationSettings;
+import com.intellij.execution.configurations.RunConfiguration;
+import com.intellij.execution.runners.ExecutionEnvironment;
+import com.intellij.execution.runners.ExecutionEnvironmentBuilder;
+import com.intellij.openapi.actionSystem.AnAction;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.project.Project;
+import io.flutter.FlutterInitializer;
+import io.flutter.run.SdkFields;
+import io.flutter.run.SdkRunConfig;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.*;
+
+
+public abstract class FlutterRunModeAction extends AnAction {
+
+  private final @NotNull String myRunArg;
+  private final @NotNull String myExecutorId;
+
+  public FlutterRunModeAction(@NotNull String text,
+                              @NotNull String description,
+                              @NotNull Icon icon,
+                              @NotNull String runArg,
+                              @NotNull String executorId) {
+    super(text, description, icon);
+    myRunArg = runArg;
+    myExecutorId = executorId;
+  }
+
+  @Override
+  public void actionPerformed(AnActionEvent e) {
+    FlutterInitializer.sendAnalyticsAction(this);
+
+    final Project project = e.getProject();
+    if (project == null) {
+      return;
+    }
+
+    final RunnerAndConfigurationSettings settings = RunManagerEx.getInstanceEx(project).getSelectedConfiguration();
+    if (settings == null) {
+      return;
+    }
+
+    final RunConfiguration configuration = settings.getConfiguration();
+    if (!(configuration instanceof SdkRunConfig)) {
+      //TODO: supported in Bazel?
+      return;
+    }
+
+    final SdkRunConfig sdkRunConfig = (SdkRunConfig)configuration.clone();
```

Comment:
It seems like cloning the run config will defeat the check that makes sure you don't launch the same config multiple times. Instead how about using putUserData() on the ExecutionEnvironment?